pkg/apis/config/v1alpha1/types.go

```diff
@@ -0,0 +1,276 @@
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Policy balloons.Config `json:"policy"`
+	// +optional
+	Common CommonConfig `json:"common,omitempty"`
+	Status ConfigStatus `json:"status,omitempty"`
+}
+
+// BalloonsPolicyList represents a list of BalloonsPolicies.
+// +kubebuilder:object:root=true
+type BalloonsPolicyList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	Items []BalloonsPolicy `json:"items"`
+}
+
+// TemplatePolicy represents the configuration for the template policy.
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +genclient
+type TemplatePolicy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Policy template.Config `json:"policy"`
+	// +optional
+	Common CommonConfig `json:"common,omitempty"`
+	Status ConfigStatus `json:"status,omitempty"`
+}
+
+// TemplatePolicyList represents a list of TemplatePolicies.
+// +kubebuilder:object:root=true
+type TemplatePolicyList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	Items []TemplatePolicy `json:"items"`
+}
+
+type CommonConfig struct {
+	// +optional
+	Control control.Config `json:"control,omitempty"`
+	// +optional
+	Log log.Config `json:"log,omitempty"`
+	// +optional
+	Instrumentation instrumentation.Config `json:"instrumentation,omitempty"`
+}
+
+// ConfigStatus is the per-node status for a configuration resource.
+type ConfigStatus struct {
+	Nodes map[string]NodeStatus `json:"nodes"`
+}
+
+// NodeStatus is the configuration status for a single node.
+type NodeStatus struct {
+	// Status of activating the configuration on this node.
+	// +kubebuilder:validation:Enum=Success;Failure
+	Status string `json:"status"`
+	// Generation is the generation the configuration this status was set for.
+	Generation int64 `json:"generation"`
+	// Error can provide further details of a configuration error.
+	Error *string `json:"errors,omitempty"`
+	// Timestamp of setting this status.
+	Timestamp metav1.Time `json:"timestamp,omitempty"`
+}
+
+const (
+	StatusSuccess = metav1.StatusSuccess
+	StatusFailure = metav1.StatusFailure
+)
+
+// NewNodeStatus returns a node status for the given error.
+func NewNodeStatus(err error, generation int64) *NodeStatus {
+	s := &NodeStatus{
+		Generation: generation,
+		Timestamp:  metav1.Now(),
+	}
+	if err == nil {
+		s.Status = StatusSuccess
+		// TODO(klihub): 'Patch away' any old errors from lingering. I don't
+		//     know if there is a nicer way of doing this with Patch().
+		e := ""
+		s.Error = &e
+	} else {
+		s.Status = StatusFailure
+		e := fmt.Sprintf("%v", err)
+		s.Error = &e
+	}
+	return s
+}
+
+type patchConfig struct {
+	Status patchStatus `json:"status,omitempty"`
+}
+
+type patchStatus struct {
+	Nodes map[string]*NodeStatus `json:"nodes,omitempty"`
+}
+
+func NodeStatusPatch(node string, status *NodeStatus) ([]byte, types.PatchType, error) {
```

Comment:
I'd suggest to keep the `types.go` strictly for typedefs. Put funcs somewhere else, preferably in a separate package (really separating the API from implementation details).

Again, this is a code-structuring thing that can be addressed later.

Reply:
Generally I agree, but I still prefer to keep code with the types when the implementation details are tightly coupled with the actual (structure of the) configuration data. And this is such a case.

Reply:
We can debate over this at some other time 😅 Even in this case I think it would be good to keep the api package cleaner. In general keeping all implementation somewhere else e.g. prevents the automatic code generation from accidentally picking up types that are not part of the API etc.

But as said, this can be addressed later if desired. In any case, I'd suggest to move the helper/accessor funcs in a separate file to keep types.go tidy. WDYT?

Reply:
But I will split all but the types to separate files in the package. I agree that it is cleaner that way.

Reply:
Updated, splitting out any extra code from the type definitions.